runtime/rasdump/trigger.c

```diff
@@ -555,7 +556,53 @@ prepareForDump(struct J9JavaVM *vm, struct J9RASdumpAgent *agent, struct J9RASdu
 		newState |= J9RAS_DUMP_TRACE_DISABLED;
 	}
 
-	if ((context->eventFlags & (J9RAS_DUMP_ON_GP_FAULT | J9RAS_DUMP_ON_ABORT_SIGNAL | J9RAS_DUMP_ON_TRACE_ASSERT)) == 0) {
+	/* Release vm access until this thread has the dumpKey and is ready to run. This will allow other threads to obtain exclusiveVMAccess in the meantime. */
+	if (vmThread) {
+		if (vmThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS) {
```

Comment:
This is closer but we still need to take `isNative` into account as I believe we have vmaccess but be in native.  @gacholio

Reply:
I'll take a detailed look before the end of the week.

Reply:
There's code in here that's already assuming we're in the JNI context when this function is running (calls AttachCurrentThreadAsDaemon) - what is the actual context when this is called? There are 3 possibilities:

- inNative (i.e. JNI)
- not inNative, have VM access
- not inNative, don't have VM access

Reply:
ah. I see now the difference with inNative is with atomic_free_jni. I will update that case.

Reply:
updated.

Reply:
I'm wrong about the attach case - that's for a signal on a thread not attached to the VM at all (so not the typical "JNI context"). The code has to handle atomic-free in the same way as below - detect and save the state so it can be restored later.